### DISCUSSION TITLE
Allow "@" character as a prefix to enable keywords to be used as identifiers.

### DISCUSSION THREAD

**Discussion by Hyrtwol** [odin-lang/Odin#3510] [Ideas/Requests]
Idea from C#
Allow "@" character as a prefix to enable keywords to be used as identifiers.
https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/lexical-structure#643-identifiers
e.g. keywords like
```
@packages
@context
@map
```
Useful for json serialization so you don't need to add attributes to map the fields.
C# example
```csharp
    [Fact]
    public void KeywordsAsIdentifiers()
    {
        int @int = 666;
        Assert.Equal(666, @int);
        Assert.Equal("int", nameof(@int));
    }
```
not sure if using @ will clash with the Odin syntax
```
@(...)
```
if so another char could be chosen ofc

**Comment by JesseRMeyer** [comment]
`int : int = 0` is already valid Odin syntax.  So at least shadowing of type names is already a feature.  I'd be shocked if shadowing of keywords more generally becomes supported.

**Reply by Hyrtwol** [reply]
Oh didn't know `int : int = 0` was allowed but that looks like a tough debugging session if you're not careful ;)
Tbh I would also be shocked to see this in Odin but you never know, maybe down the road. Can't remember exactly when it was added to C# but pretty sure it was not in version 1.0 :p
Also I don't think you can do that trick with all keywords, this don't work
```
map : int = 0
context : int = 0
```

**Comment by perryhq** [comment]
Can't you use another character like `_`?

**Reply by flysand7** [reply]
`_` is a black-hole variable, it consumes everything, so it is a special name which you can't declare too.

**Reply by perryhq** [reply]
I meant prefixing the identifier with `_`

**Comment by flysand7** [comment]
I'm pretty sure our JSON marshaller supports alt-names for fields and it can understand when a field in a struct differs from the encoded field. That said, serialization is the only place where this syntax is useful. If you were fine with `@foo` being special, maybe this isn't too bad:


```
JSON_Entity :: struct {
   name: string,
   value: f64,
   _context: string, `json:"context"`
}
```

To specify a different encoded field name. The benefit of this feature doesn't extend beyond the problem of serialization and can't be abused like the `@kw` thing.

**Comment by gingerBill** [comment]
I know this is an old post but there is zero need for this whatsoever. Why do you need to use a keyword as an identifier? Every place where I think people would think it is necessary is the following:

* `foreign` procedures/variables. `@(link_name=<string>)` exists so that the link name and the usage name can be separate
* Struct fields. The struct field name doesn't have to match some `foreign` code, and if you need it to be different for serialization/marshalling purposes, use struct field tags.
* You just want to use that work. Just use a different one. `context` is usually written with `ctx`, or even `_context` or `context_`.

**Reply by jqcorreia** [reply]
Thanks for the clarification here @gingerBill . I'm in the middle of writing a barebones wrapper for libwayland (that I fully intend to submit)  and I stumbled upon a struct with a `in` field (which is obviously a keyword). But a doubt persists: How come the name doesn't matter on a `foreign` code import?

**Reply by perryhq** [reply]
I think he meant that you can use `@(link_name)` to link correctly while you use a different name like `_in` inside your odin code.
